### ISSUE TITLE
[bug] `blocks` will be reset on `preprocess` every time and will break `.md` in `.md`

### ISSUE THREAD

**Issue by JounQin** [eslint/eslint-plugin-markdown#181]
The  will not work when writing markdown codes in code blocks.

Lifecycle:

```
preprocess('.md') -> preprocess('.md/0_0.md') -> postprocess('.md/0_0.md') -> postprocess('.md')
```

This was not a problem before because `.md` files are not linted by eslint before. By adding `esling-plugin-mdx` which support `.md` files, it will be a problem.

_Originally posted by @JounQin in https://github.com/eslint/eslint-plugin-markdown/pull/178#discussion_r593347558_

---

@btmills 

It's clear to be an bug

https://github.com/eslint/eslint-plugin-markdown/blob/main/lib/processor.js#L219

**Comment by btmills** [comment]
Thanks for opening this so we can dig in. I think I follow what you're saying, but to be sure, can you provide an example Markdown document that triggers the bug?

**Comment by JounQin** [comment]
@btmills Here you are.

https://repl.it/@JounQin/eslint-plugin-markdown-181#test.md

![image](https://user-images.githubusercontent.com/8336744/111584735-37573280-87f9-11eb-9433-3630ee8a4478.png)

4 warning is expected, two from top level file, two from code block which are not reported.

**Comment by JounQin** [comment]
This issue will only occur when integrating with other plugins like `eslint-plugin-mdx` which will lint original `.md` files by `preprocess` too.

Here is the fake version for testing.

https://github.com/eslint/eslint-plugin-markdown/pull/183/files#diff-83e0a2056017ee834945f51308072e4a883a4ed96193581fe4aabc41cb98710a

**Comment by JounQin** [comment]
So maybe I shouldn't flag it as bug but a better compatibility with other markdown related ESLint plugins based on `eslint-plugin-markdown`.

**Comment by btmills** [comment]
This seems related to https://github.com/eslint/eslint/pull/14227/files#r602802758. I can reproduce this today in a test if I configure this plugin to run on `.md` and `.markdown` files and have a `README.md` file containing a `markdown` code block. If #14227 changes ESLint to run processors directly recursively, this would be even easier to reproduce.

**Comment by JounQin** [comment]
Yes and no.

This issue is not always caused by same extension, `markdown` in `md` should also result this issue.

**Comment by JounQin** [comment]
Yes and no.

This issue is not always caused by same extension, `markdown` in `md` should also result this issue.

**Comment by JounQin** [comment]
Yes, I should simplify the test case in current PR.
